### DISCUSSION TITLE
[Dev] Better type annotations

### DISCUSSION THREAD

**Discussion by DanielYang59** [materialsproject/pymatgen#3724] [General]
As we're pushing towards more (static) type annotation coverage across the code base, I'm experiencing multiple issues and I would like to hear about your suggestions on the best practice here.

1. Use of "functionally similar" data types, for example as discussed [here](https://github.com/materialsproject/pymatgen/pull/3705#discussion_r1542383319): 
For example the following `Kpoint` class annotated the `kpts_shift` arg as `Vector3D`(`tuple[float, float, float]`): https://github.com/materialsproject/pymatgen/blob/25006642a53bab0ac78c5415d21ec881e767dec2/pymatgen/io/vasp/inputs.py#L1024-L1042

As a result codes that returns a `list[float]` would be flagged as having wrong types, for example multiple lines in the following that generate a `list[float]` instead of `tuple[float, float, float]`: https://github.com/materialsproject/pymatgen/blob/25006642a53bab0ac78c5415d21ec881e767dec2/pymatgen/io/vasp/inputs.py#L1371-L1479

2. Handling of default values (especially `None`), for example in 3b293bf303028e92786de9c6b7e74e8fb57ed744 and https://github.com/materialsproject/pymatgen/blob/25006642a53bab0ac78c5415d21ec881e767dec2/pymatgen/io/vasp/optics.py#L139-L193. This would require explicit check for `None`, which would surely make the code more robust (but more "messy" as well).

Any advice would be appreciated. @janosh

**Comment by janosh** [comment]
sorry for the lack of reply, i completely missed this. are there any open questions left here?
